Fixed #37 -- Refactor bitboard play and play_pvp

amcintosh opened the pull request:
Refactor to remove duplicate code between the two.

Fixes #37

amcintosh committed:
Fixed #37 -- Refactor bitboard play and play_pvp

Refactor to remove duplicate code between the two.

hrushikeshrv commented on chessengine/bitboard.py:
```diff
@@ -609,6 +611,49 @@ def alpha_beta_search(
+                try:
+                    self.undo_move()
+                    if not is_pvp:
+                        self.undo_move()  # If computer has moved, we need to undo it too
```
On second thought, shouldn't we execute this statement even if 2 human players are playing against each other? If it is player 1's turn and they perform an undo, that would undo the previous move, which was actually player 2's move. To undo player 1's move and return control to them, we would have to undo another move.

amcintosh replied:
Good point. I'll update that. 

Also I see black isn't happy with some pre-existing code. I'll run it and fix that up.

amcintosh replied:
Done

amcintosh committed:
Fixed hrushikeshrv#37 -- Revert both players moves in pvp

Instead of just against the computer. This means a player making an undo
always rolls back to their last move.

Also includes black formatting changes

hrushikeshrv commented on chessengine/bitboard.py:
```diff
@@ -617,6 +606,48 @@ def alpha_beta_search(
+        """Wrapper for input to allow testing"""
+        return input(prompt).strip()
+
+    def handle_player_move(self, side_to_move: str, last_move: str) -> Tuple[str, str, bool]:
```
The return type for `handle_player_move` should be `Tuple[str, int, bool]`

hrushikeshrv commented on chessengine/bitboard.py:
```diff
@@ -724,34 +727,8 @@ def play_pvp(self) -> None:
-                        print("No moves have been made yet to undo!")
-                        lines_printed += 1
-                    break
+            move, lines_added, _ = self.handle_player_move(side_to_move, last_move)
```
Here we should add a check to see if a move was undone and continue if so, just like in `play` -
```python
if move_undone:
    continue
```

amcintosh replied:
Ah yes. That would be required with the pvp undo change.

hrushikeshrv commented on chessengine/bitboard.py:
```diff
@@ -617,6 +606,48 @@ def alpha_beta_search(
 
+    def get_input(self, prompt: str) -> str:
+        """Wrapper for input to allow testing"""
+        return input(prompt).strip()
```
I'm not sure about the purpose of this wrapper. Do you plan to extend it's functionality? I think we could do without it since it doesn't do much at the moment.

amcintosh replied:
This was entirely for test mocking purposes (https://github.com/amcintosh/chessengine/blob/issue-37-refactor-play-play-pvp/tests/test_bitboard.py#L76-L78). `input()` cannot be mocked directly, but the wrapper can.

hrushikeshrv replied:
Oh, I see. Everything looks good then, could you run `black` once on your changes to get the linting to pass? Then I'll be able to merge.

amcintosh replied:
Updated!

amcintosh replied:
Not really sure why it's failing. Black seesm happy locally:

```shell
$ black chessengine/bitboard.py 
All done! ✨ 🍰 ✨
1 file would be left unchanged.

$ black tests/test_bitboard.py 
All done! ✨ 🍰 ✨
1 file would be left unchanged.
```

hrushikeshrv replied:
Can you try updating your `black` version once? I just upgraded to version `22.10.0` locally, and it does tell me to reformat both those files. Once this is done, I think I'll pin `black`'s version to `22.10.0` in the GitHub action for consistency since `black` seems to do this often.

amcintosh replied:
Ah. That did it. Thanks

amcintosh committed:
Fixed hrushikeshrv#37 -- Cleanup

- Fixed method return signature
- Change to pvp undo (now undoes both player's moves) handled in
  play_pvp
- Fix up tests

amcintosh committed:
Lint fixes from latest black version

hrushikeshrv commented:
Thanks for this! 👍🏻
